l1-contracts/contracts/bridgehub/Bridgehub.sol

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+import "./IBridgehub.sol";
+import "../bridge/interfaces/IL1SharedBridge.sol";
+import "../state-transition/IStateTransitionManager.sol";
+import "../common/ReentrancyGuard.sol";
+import "../state-transition/chain-interfaces/IZkSyncStateTransition.sol";
+import {ETH_TOKEN_ADDRESS, TWO_BRIDGES_MAGIC_VALUE} from "../common/Config.sol";
+import {BridgehubL2TransactionRequest} from "../common/Messaging.sol";
+import "../vendor/AddressAliasHelper.sol";
+
+contract Bridgehub is IBridgehub, ReentrancyGuard, Ownable2Step {
+    /// @notice all the ether is held by the weth bridge
+    IL1SharedBridge public sharedBridge;
+
+    /// @notice we store registered stateTransitionManagers
+    mapping(address _stateTransitionManager => bool) public stateTransitionManagerIsRegistered;
+    /// @notice we store registered tokens (for arbitrary base token)
+    mapping(address _token => bool) public tokenIsRegistered;
+
+    /// @notice chainID => StateTransitionManager contract address, storing StateTransitionManager
+    mapping(uint256 _chainId => address) public stateTransitionManager;
+
+    /// @notice chainID => baseToken contract address, storing baseToken
+    mapping(uint256 _chainId => address) public baseToken;
+
+    /// @notice gasPriceMultiplier for each baseToken, so that each L1->L2 transaction pays for its transaction on the destination
+    mapping(address _baseToken => uint256) public baseTokenGasPriceMinMultiplier;
```

Comment:
Note that the fact that the multiplier is an integer means that it will not work for tokens that have fewer digits than ETH. For instance USDC has 6 decimals. 

So a price of 50 gwei will lead to 50000$ per gas.

Reply:
Even if you decide to "normalize" based on decimals, still tokens like wBTC wont work for now (maybe it is fine for teh initial release though)

Reply:
Why not just use fixed point numbers?

Reply:
Because it has higher price than ETH, and so if the numbers are kept as integers, you can not multiply it by less 1. 

Right now it is calculated `gas_price_in_eth * K, K >= 1`, in case of wBTC it would lead to severe overestimation of cost.

Yes, fixed point arithmetic would do the trick